qiskit/circuit/library/boolean_logic.py

```diff
@@ -0,0 +1,118 @@
+"""
+
+from typing import List, Optional
+
+import numpy as np
+from qiskit.circuit import QuantumRegister, QuantumCircuit
+from qiskit.circuit.exceptions import CircuitError
+
+
+class Permutation(QuantumCircuit):
+    """An n_qubit circuit that permutes qubits."""
+
+    def __init__(self,
+                 n_qubits: int,
+                 pattern: Optional[List[int]] = None,
+                 seed: Optional[int] = None) -> QuantumCircuit:
+        """Return an n_qubit permutation circuit impelemented using SWAPs.
+
+        Args:
+            n_qubits: circuit width.
+            pattern: permutation pattern. If None, permute randomly.
+            seed: random seed in case a random permutation is requested.
+
+        Returns:
+            A permutation circuit.
+
+        Raises:
+            CircuitError: if permutation pattern is malformed.
+        """
+        super().__init__(n_qubits, name="permutation")
+
+        if pattern is not None:
+            if sorted(pattern) != list(range(n_qubits)):
+                raise CircuitError("Permutation pattern must be some "
+                                   "ordering of 0..n_qubits-1 in a list.")
+            pattern = np.array(pattern)
+        else:
+            rng = np.random.RandomState(seed)
+            pattern = np.arange(n_qubits)
+            rng.shuffle(pattern)
+
+        for i in range(n_qubits):
+            if (pattern[i] != -1) and (pattern[i] != i):
+                self.swap(i, int(pattern[i]))
+                pattern[pattern[i]] = -1
+
+
+class Shift(QuantumCircuit):
+    """An n_qubit circuit that shifts inputs over Z2."""
+
+    def __init__(self,
+                 n_qubits: int,
+                 amount: Optional[int] = None,
+                 seed: Optional[int] = None) -> QuantumCircuit:
+        """Return a circuit implementing bitwise xor (shift over Z_2).
+
+        Args:
+            n_qubits: the width of circuit.
+            amount: shift amount in decimal form.
+            seed: random seed in case a random shift amount is requested.
+
+        Returns:
+            A boolean shift circuit.
+
+        Raises:
+            CircuitError: if the shift amount exceeds available qubits.
+        """
+        super().__init__(n_qubits, name="shift")
+
+        if amount is not None:
+            if len(bin(amount)[2:]) > n_qubits:
+                raise CircuitError("Bits in 'amount' exceed circuit width")
+        else:
+            rng = np.random.RandomState(seed)
+            amount = rng.randint(0, 2**n_qubits)
+
+        for i in range(n_qubits):
+            bit = amount & 1
+            amount = amount >> 1
+            if bit == 1:
+                self.x(i)
+
+
+class InnerProduct(QuantumCircuit):
+    """An n_qubit circuit that computes the inner product of two registers."""
+
+    def __init__(self, n_qubits: int) -> QuantumCircuit:
+        """Return a circuit to compute the inner product of 2 n-qubit registers.
+
```

Comment:
I think it'd help to add a note that the inner product is realized via CZ gates. Would it be useful to allow for different gates?

Reply:
sounds good.